### PR TITLE
Add llvm 18 support

### PR DESCRIPTION
**Pull request by mcbarton**
This PR adds Clang-18 to the CI now that it has been released.

### CODE REVIEW

**Comment by mcbarton** [comment]
@vgvassilev This PR is ready for review.

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,11 +56,15 @@ if( CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR )
   ## Define supported version of clang and llvm
 
   set(CLANG_MIN_SUPPORTED 13.0)
-  set(CLANG_MAX_SUPPORTED "17.0.x")
-  set(CLANG_VERSION_UPPER_BOUND 17.1.0)
+  set(CLANG_MAX_SUPPORTED "18.1.x")
+  #FIXME: As 19.0.0 will not exist due to new naming convention this sets the upper bound to any version of clang 18
```
Comment (vgvassilev) [inline CMakeLists.txt:60]: I am not sure if I understand what that FIXME is about.

**Review by vgvassilev — COMMENTED** [review]
Can you run `git-clang-format HEAD~` to make the bot happy?

**Comment by mcbarton** [comment]
@vgvassilev It is passing all checks now.

**Comment by mcbarton** [comment]
@alexander-penev can you review this PR for me please?

**Comment by vgvassilev** [comment]
We will need to drop support for clang7. Our policy is to support the last 10 versions. That would require to clean up the `ifdef`s that mention clang 7.

**Comment by vgvassilev** [comment]
Can you edit the commit message to something like:
```
Add support for clang18.

Fixes #805.
```

**Comment by mcbarton** [comment]
> Can you edit the commit message to something like:
> 
> ```
> Add support for clang18.
> 
> Fixes #805.
> ```

I can change the commit message to "Add support for clang18" . We have no specific issue of support for clang 18 at the moment. I can create one and then add the fix message as you suggest.

**Comment by mcbarton** [comment]
> We will need to drop support for clang7. Our policy is to support the last 10 versions. That would require to clean up the `ifdef`s that mention clang 7.

I will look for any ifdefs that mention clang 7 tomorrow and remove them.

**Comment by vgvassilev** [comment]
> > Can you edit the commit message to something like:
> > ```
> > Add support for clang18.
> > 
> > Fixes #805.
> > ```
> 
> I can change the commit message to "Add support for clang18" . We have no specific issue of support for clang 18 at the moment. I can create one and then add the fix message as you suggest.

Dammit, no, I thought this is on the clad repository :(

**Comment by vgvassilev** [comment]
> > We will need to drop support for clang7. Our policy is to support the last 10 versions. That would require to clean up the `ifdef`s that mention clang 7.
> 
> I will look for any ifdefs that mention clang 7 tomorrow and remove them.

Please ignore that comment...

**Comment by mcbarton** [comment]
@vgvassilev @alexander-penev after this PR is reviewed and merged I am going to add a xeus-cpp build to the end of the wasm CppInterOp CI jobs. This will allow me/you to test possible changes to get the correct extension in https://github.com/compiler-research/xeus-cpp/pull/14#issuecomment-1989465806

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**